lib/chat/key_share.ex

```diff
@@ -53,6 +53,54 @@ defmodule Chat.KeyShare do
     end)
   end
 
+  def compose([], upload_shares), do: MapSet.new(upload_shares)
+
+  def compose(shares, upload_shares) do
+    MapSet.union(MapSet.new(shares), MapSet.new(upload_shares))
+  end
+
+  def look_for_duplicates(shares) do
+    shares
+    |> Enum.filter(fn share -> Enum.count(shares, &(&1.key == share.key)) > 1 end)
+    |> Enum.group_by(& &1.key)
+    |> Enum.map(fn {key, maps} ->
+      %{
+        key: key,
+        exclude: maps |> Enum.min_by(&(&1.ref |> String.to_integer())) |> Map.get(:ref),
+        ref: Enum.map(maps, & &1.ref)
+      }
+    end)
+  end
+
+  def read_content(path) do
+    content =
+      path
+      |> File.stream!()
+      |> Stream.map(&String.trim_trailing/1)
+      |> Enum.to_list()
+      |> List.to_tuple()
```

Comment:
insert before this line
`|> Enum.map(&decode_content/1)`
and no need to make final tuple

Reply:
right